src/main/java/edu/uwaterloo/ece658/service/BrowseService.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package edu.uwaterloo.ece658.service;
+
+import edu.uwaterloo.ece658.entity.Photo;
+import edu.uwaterloo.ece658.entity.Tag;
+import edu.uwaterloo.ece658.entity.User;
+import edu.uwaterloo.ece658.session.TagFacade;
+import edu.uwaterloo.ece658.session.UserFacade;
+import edu.uwaterloo.ece658.session.UserTagFacade;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.ejb.EJB;
+import javax.ejb.Stateless;
+
+/**
+ * Service that handles all kinds of browsing requests from front end.
+ *
+ * @author mier
+ */
+@Stateless
+public class BrowseService {
+
+    @EJB
+    private UserFacade userFacade;
+    @EJB
+    private TagFacade tagFacade;
+    @EJB
+    private UserTagFacade userTagFacade;
+
+    /**
+     * Given a username, fetch all the photos uploaded by this user.
+     *
+     * @param userName user name whose uploaded photos are being fetched
+     * @return list of photos uploaded by this user
+     */
+    public List<Photo> fetchUploadedPhotosOfUser(String userName) {
+        User user = userFacade.retrieveUserByUserName(userName);
+        return user.getUploadedPhotos();
+    }
+
+    /**
+     * Given a username, fetch all photos classified by the subscribed tags by
+     * this user.
+     *
+     * @param userName user name whose uploaded photos are being fetched
+     * @return mapping between subscribed tag and photos under the tag of the
+     * user
+     */
+    public Map<Tag, List<Photo>> fetchPhotosClassifiedBySubscribedTagsOfUser(String userName) {
+        User user = userFacade.retrieveUserByUserName(userName);
+        Map<Tag, List<Photo>> result = new HashMap<>();
+        for (Tag tag : user.getSubscribedTags()) {
+            // User explicitly subscribed to some tags => Directly show photos
+            // under these tags is ok. This includes UserTag that the user
+            // has subscribed to.
+            result.put(tag, tag.getPhotosUnderThisTag());
```

Comment:
As an afterthought.... Where do we deal with private photos? As a private photo still may be under one of these tags. Will the front end not show these? Or should they be filtered out before passing the Map

Reply:
Good question.
The point of private photo is only to say that "please don't notify my private photo to any one that subscribed to the tags of my private photo". notifyXXX() methods can expose private methods so we need to have a check there. Here this method can still leak the private photos. So I need to have a check, too.
Doing it at the backend should be correct. Even though you front-end doesn't show private photos, client can still find a way to get private photos if only your backend returns them.